src/adaXT/decision_tree/decision_tree.pyx

```diff
@@ -175,52 +180,71 @@ class DecisionTree(BaseModel):
             self.__check_dimensions(X)
         return np.asarray(self.predictor.predict(X, **kwargs))
 
-    def predict_proba(self, X: ArrayLike) -> np.ndarray:
-        if not self.predictor:
-            raise AttributeError("The tree has not been fitted before trying to call predict_proba")
-        if not self.skip_check_input:
-            X, _ = self.__check_input(X)
-            self.__check_dimensions(X)
-        return np.asarray(self.predictor.predict_proba(X))
-
-    def __get_leaf_matrix(self, scale: bool = False) -> np.ndarray:
-        cdef:
-            int i, j
-            int[::1] indices
-            int n_node
-            int n_rows
-
+    def __get_leaf(self, scale: bool = False) -> dict:
         if not self.root:
             raise ValueError("The tree has not been trained before trying to predict")
 
         leaf_nodes = self.leaf_nodes
         if (not leaf_nodes):  # make sure that there are calculated observations
             raise ValueError("The tree has no leaf nodes")
 
-        n_rows = self.n_rows
-        matrix = np.zeros((n_rows, n_rows))
+        ht = {}
         for node in leaf_nodes:
-            indices = node.indices
-            for i in indices:
-                for j in indices:
-                    if scale:
-                        n_node = len(node.indices)
-                        matrix[i, j] += 1/n_node
-                    else:
-                        matrix[i, j] += 1
+            ht[node.id] = node.indices
+        return ht
+
+    def _tree_based_weights(self, hash1: dict, hash2: dict, size_X0: int,
+                            size_X1: int, scale: bool) -> np.ndarray:
+        matrix = np.zeros((size_X0, size_X1))
+        hash1_keys = hash1.keys()
+        hash2_keys = hash2.keys()
+        for xi in hash1_keys:
+            if xi in hash2_keys:
+                indices_1 = hash1[xi]
+                indices_2 = hash2[xi]
+                if scale == 0:
```

Comment:
I see how the 3 different scalings are needed, but reading it over now it just seems confusing. Perhaps we should instead have scaling be a string, and then come up with 3 different names, such that it is more easily distinguishable.

Reply:
Good idea, how about "no_scale", "column_scale" and "symmetric_scale"?

Reply:
Or better: "none", "column" and "symmetric" (scaling is unnecessary).